dali/core/mm/cuda_vm_resource_test.cc

```diff
@@ -0,0 +1,305 @@
+    void *ptr4 = res.allocate(150<<20);
+    EXPECT_EQ(ptr4, static_cast<void*>(static_cast<char*>(ptr2) + (100<<20)));
+    for (int i = 0; i < 3; i++) {
+      // ptr4 should start 200 MiB from start, which is block 6
+      // block 7 should be still unmapped, hence i+7.
+      // Let's check that the 1st 3 blocks have been reused rather than newly allocated.
+      EXPECT_EQ(res.va_regions_.back().mapping[i+7], blocks[i]);
+    }
+    res.deallocate(ptr2, 100<<20);
+    res.deallocate(ptr4, 150<<20);
+  }
+
+  void MapRandomBlocks(cuda_vm_resource::va_region &region, int blocks_to_map) {
+    assert(blocks_to_map < region.num_blocks());
+    std::vector<int> block_idxs(region.num_blocks());
+    random_permutation(block_idxs, rng_);
+    block_idxs.resize(blocks_to_map);
+    for (int blk_idx : block_idxs) {
+      region.map_block(blk_idx, cuvm::CUMem::Create(region.block_size));
+    }
+  }
+
+  // Unlinke normal va_region, this one doesn't release the memory blocks upon destruction
+  struct va_region_backup : cuda_vm_resource::va_region {
+    using va_region::va_region;
+    va_region_backup(va_region_backup &&other) = default;
+
+    ~va_region_backup() {
+      // suppress purge
+      mapping.clear();
+    }
+  };
+
+  static va_region_backup Backup(const cuda_vm_resource::va_region &in) {
+    va_region_backup out(in.address_range, in.block_size);
+    out.available_blocks = in.available_blocks;
+    out.mapping   = in.mapping;
+    out.mapped    = in.mapped;
+    out.available = in.available;
+    return out;
+  }
+
+  void ComparePart(const cuda_vm_resource::va_region &region,
+                   int pos,
+                   const cuda_vm_resource::va_region &ref) {
+    for (int i = 0, j = pos; i < ref.num_blocks(); i++, j++) {
+      EXPECT_EQ(region.mapping[j],   ref.mapping[i])   << "@ " << j;
+      EXPECT_EQ(region.mapped[j],    ref.mapped[i])    << "@ " << j;
+      EXPECT_EQ(region.available[j], ref.available[i]) << "@ " << j;
+    }
+  }
+
+  void TestRegionExtendAfter() {
+    cuda_vm_resource res;
+    res.block_size_ = 4 << 20;  // 4 MiB
+    const int b1 = 32, b2 = 64;
+    const size_t s1 = b1 * res.block_size_;
+    const size_t s2 = b2 * res.block_size_;
+    res.va_ranges_.push_back(cuvm::CUMemAddressRange::Reserve(s1 + s2));
+    cuvm::CUAddressRange total = res.va_ranges_.back();
+    cuvm::CUAddressRange part1 = { total.ptr(),           s1 };
+    cuvm::CUAddressRange part2 = { total.ptr() + s1,      s2 };
+    res.va_add_region(part1);
+    ASSERT_EQ(res.va_regions_.size(), 1u);
+    MapRandomBlocks(res.va_regions_[0], 10);
+    va_region_backup va1 = Backup(res.va_regions_[0]);
+    res.va_add_region(part2);
+    ASSERT_EQ(res.va_regions_.size(), 1u);
+    auto &region = res.va_regions_.back();
+    ASSERT_EQ(region.num_blocks(), b1 + b2);
+    ComparePart(region, 0, va1);
+    for (int i = b1; i < b1 + b2; i++) {
+      EXPECT_FALSE(region.mapped[i]);
```

Comment:
Is it because `MapRandomBlocks` was not called on the `part2`?

Reply:
Yes, it should be free but unmapped VA.